gh-84644: Fix singledispatch annotation parsing (#143465)
python/cpython: Misc/NEWS.d/next/Library/2026-01-06-09-13-53.gh-issue-84644.V_cYP3.rst

:func:`functools.singledispatch` and :func:`functools.singledispatchmethod`
now require callables to be correctly annotated if registering without a type explicitly
specified in the decorator. The first user-specified positional parameter of a callable

> must always be annotated. Before, a callable could be registered based on its return type

Reviewer: I'm not sure I catch the two first sentences correctly. Is it equivalent to:

```rst
Using :func:`functools.singledispatch` and :func:`functools.singledispatchmethod`
without specifying the dispatching type now require the first positional parameter
of the decorated callables to be annotated.
```

Author: In general yes, with an asterisk that "the first positional parameter" would mean `arg` -- not `self` -- in this example method `C.f`:

```py
class C:
    def f(self, arg):
		...
```

Maybe this can be rephrased entirely.

I'll go with:

```suggestion
When registering a callable to :func:`functools.singledispatch`
or :func:`functools.singledispatchmethod` via a type annotation,
it is now validated that the first proper positional parameter is annotated.
Before, a callable could be registered based on its return type
```

Reviewer: "Proper" is either too vague or could be misunderstood. Maybe "first positional parameter" for singledispatch and first non-self positional parameter for singledispatchmethod?

Author: Yes, we already use "non-*self* or non-*cls*" in our documentation. I'll use it here as well.

Author: I've chosen not to go into details on how singledispatch/singledispatchmethod particularly affect the semantics.

> annotation or based on an irrelevant parameter type annotation. Contributed by Bartosz Sławecki.

Reviewer: ```suggestion
annotation or based on an irrelevant parameter type annotation. Patch by Bartosz Sławecki.
```

I know there is no rule here but most (if not all) NEWS entries are "Patch by XXX" while What's News ones are "Contributed by XXX in YYY". So let's keep it that way.

Author: Interesting, I haven't known that and used to flip back and forth between the two.

> annotation or based on an irrelevant parameter type annotation. Contributed by Bartosz Sławecki.

Reviewer: ```suggestion
must always be annotated. Previously, a callable could be registered based on its return type
annotation or an irrelevant parameter type annotation. Contributed by Bartosz Sławecki.
```
